### ISSUE TITLE
Controller accessRules evaluated before 'req.auth.ajax.user'

### ISSUE THREAD

**Issue by tothand** [evan108108/RESTFullYii#99]
Hi,

I'm fairly new to Yii, so maybe I misunderstood something.

In the config/main.php I've created a section like this:

``` php
    'params'=>array(
        'RestfullYii' => array(
            'req.auth.ajax.user' => function(){
                if(isset($_SERVER['HTTP_X_REST_USERNAME']) and isset($_SERVER['HTTP_X_REST_PASSWORD'])) {
                    $username = trim($_SERVER['HTTP_X_REST_USERNAME']);
                    $password = trim($_SERVER['HTTP_X_REST_PASSWORD']);
                    $identity=new UserIdentity($username,$password);
                    if($identity->authenticate()){
                        Yii::app()->user->login($identity,0);
                        return true;
                    }
                    else{
                        return false;
                    }
                }
                return false;
            },
        ),      
```

If I have a controller where the accessRules() function looks like this:

``` php
        return array(
            array('allow',
                'actions'=>array('REST.GET','REST.PUT', 'REST.POST', 'REST.DELETE'),
                'users'=>array('*'),
            ),
        );
```

The req.auth.ajax.user takes care of the authentication. If the X_REST... headers are not present in the request or the user/password is incorrect  I got a 401 (Unauthorized) response which is perfect.

BUT

If I have the following in the accessRules() function:

``` php
        return array(
            array('allow',
                'actions'=>array('REST.PUT', 'REST.POST', 'REST.DELETE'),
                'roles'=>array('webhd.admin'),
            ),
        );
```

I'm presented with the Yii default login page.

I did some debugging and I've figured out that the Yii::app()->user->checkAccess() function is called before the 'req.auth.ajax.user' code in the config/main.php so Yii thinks I'm not logged in /which is true/ and redirects me to the login page /which is bad because it is an API and the client side is not rendered here /it's a Single Page App (SPA) written is AngularJS and resides on a different domain and server/

What I'm trying to achieve is to create a pure REST API where the API is totally unaware of the client (no server side rendered login page or redirects only HTTP calls and JSON responses) so I can use my API with a SPA (built with AngularJS) and later I can create other API consumer clients (e.g. Android APP).

First I wanted to do OAuth but I couldn't find any suitable Yii OAuth 2.0 server module where I can implement the Resource Owner Password Flow so I'd like to do some basic auth but I'd like to take advantage of the role based auth mechanism in Yii.

Am I missing something? Am I on the wrong track? Is it even possible without doing some 'magic'? :)

Any help would be appreciated.

Best,
Andras
